state/protocol/badger/mutator.go

```diff
@@ -581,44 +590,54 @@ func (m *Mutator) receiptExtend(candidate *flow.Block) error {
 	// check each receipt included in the payload for duplication
 	for _, receipt := range payload.Receipts {
 
-		// if the receipt was already included before, error
-		_, duplicated := lookup[receipt.ID()]
-		if duplicated {
-			return state.NewInvalidExtensionErrorf("payload includes duplicate receipt (%x)", receipt.ID())
+		valid, err := IsValidReceipt(receipt, lookup, forkBlocks, prevReceiptHeight)
+		if err != nil {
+			return fmt.Errorf("could not check validity of Execution Receipt %v: %w", receipt.ID(), err)
 		}
-
-		// if the receipt is not for a block on this fork, error
-		if _, ok := forkBlocks[receipt.ExecutionResult.BlockID]; !ok {
-			return state.NewInvalidExtensionErrorf("payload includes receipt for block not on fork (%x)", receipt.ExecutionResult.BlockID)
+		if !valid {
+			return state.NewInvalidExtensionErrorf("payload includes invalid receipt (%x)", receipt.ID())
 		}
 
-		// get the reference block to check expiry
-		h, err := m.state.headers.ByBlockID(receipt.ExecutionResult.BlockID)
-		if err != nil {
-			return fmt.Errorf("could not get reference block (%x): %w", receipt.ExecutionResult.BlockID, err)
-		}
+		// at this point we know that forkBlocks contains the receipt's block
+		// because otherwise the receipt wouldn't be valid and the function
+		// would have already errored-out
+		prevReceiptHeight = forkBlocks[receipt.ExecutionResult.BlockID].Height
+	}
 
-		// check whether receipt is for block at or below the sealed and
-		// finalized height
-		if h.Height <= sealedHeight {
-			// Block has either already been sealed and finalized  _or_
-			// block is a _sibling_ of a sealed and finalized block.
-			// In either case, it's an error
-			return state.NewInvalidExtensionErrorf("payload includes receipt for block below sealed height (%x) (%d/%d)",
-				receipt.ExecutionResult.BlockID,
-				h.Height,
-				sealedHeight)
-		}
+	return nil
+}
 
-		// check receipts are sorted by block height
-		if h.Height < prevReceiptHeight {
-			return state.NewInvalidExtensionError("payload receipts should be sorted by block height")
-		}
+// IsValidReceipt checks the validity of a receipt against the following
+// criteria:
+// * the receipt IS NOT in receiptLookup
+// * the receipt's block IS in blockLookup
+// * the receipt's block is higher than minHeight
+func IsValidReceipt(receipt *flow.ExecutionReceipt,
```

Comment:
Sorry, my [previous comment](https://github.com/onflow/flow-go/pull/100/files#r525818744) was probably incomplete. I did _not_ mean to suggest to move the checks https://github.com/onflow/flow-go/blob/828efdd27db8253987452135aaf5a47a9ca0dab4/state/protocol/badger/mutator.go#L620-L638 into a dedicated method (though, I am also not opposed to it either).

The method `IsValidReceipt` was supposed to represent checks that are _independent_ of where the receipt was incorporate, e.g. checks like:
* is the receipt from a valid execution node with non-zero weight (stake)?
* cryptography checks ?
* does it contain the expected number of chunks?
* ...

While we implement the checks for constructing the payload (and do not include such receipts in `payload.Receipts`), a byzantine consensus node might not adhere to this rule. Therefore, we must also check that a proposal from a different node contains only valid receipts in `payload.Receipts`. 

If the receipt is invalid, the proposing consensus node should _not_ include it in a block. Instead, it should raise a challenge. Including an invalid receipt in `payload.Receipts` is a slashable protocol violation.